README.md
# coinbase_commerce

A dart library that connects to interact with the Coinbase Commerce API. 
Enables projects to connect seamlessly to coinbase and receive payments!

For more details visit Coinbase API docs.

To start using library, you need to register on Coinbase Commerce and get your API_KEY from user settings.

Next create a Coinbase object for interacting with the API:
the object contains links to an dart Class representations of the API resources Checkout, Charge and Invoice.

Each request returns a CoinbaseResponseObject.
Every CoinbaseResponseObject matches the API response Jsons and casts them as models.

# Features
* Create Charges
* Cancel Charges
* View Charges
* List Charges
* View The Status of Charges
* Create Invoices
* View Invoices
* List Invoices
* Create Checkouts
* Delete Checkouts
* View Checkouts
* List Checkouts

See also:
- [The official Coinbase Commerce API documentation](https://commerce.coinbase.com/docs/api/).

# Usage
To create a new instance of a coinbase_commerce object you have to supply your API as generated from the coinbase commerce settings page.
```dart 
Coinbase coinbase=Coinbase('YOUR API KEY HERE',debug:true);
```
### Charges
#### To create a new charge 

```dart
ChargeObject charge = await coinbase.createCharge(
                        name: 'Coffee Charge',
                        description: '3 cups of Coffee',
                        currency: CurrencyType.usd,
                        pricingType: PricingType.fixedPrice,
                        amount: 40);
```

#### To view the status of Your previously created charge

```dart
StatusObject status = StatusCheck().checkChargeStatus(image);
```
Charges can either be (NEW, PENDING, COMPLETED, EXPIRED,
UNRESOLVED, RESOLVED, CANCELED, REFUND PENDING, REFUNDED,NONE)

In the case where it returns a status of none it implies that no record of the charge was seen.

#### To view the information concerning a particular charge

```dart
ChargeObject charge = await coinbase.viewCharge('YOUR CHARGE ID HERE');
```

#### To cancel a charge
```dart
ChargeObject charge= await coinbase.cancelCharge('YOUR CHARGE ID HERE');
```
#### To list all charges assoiciated with an account

```dart
List<ChargeObject> charge = await coinbase.retrieveAllCharges;
```
### Checkouts
#### To create a new checkout
```dart
CheckoutObject checkout = await coinbase.createCheckout(
                        description: 'description',
                        name: 'name',
                        pricingType: PricingType.noPrice,
                        currency: CurrencyType.usd);
```
#### To view the information concerning a particular checkout

```dart
CheckoutObject charge = await coinbase.viewCheckout('YOUR CHECKOUT ID HERE');
```
#### To update a previously created checkout, supply the id and the new information,
```dart
CheckoutObject checkout = await coinbase.updateCheckout(
                        checkoutID: 'c4357d83-42d4-4d7b-bc44-d0262b1efd25',
                        name: 'New Name here');
```
#### To delete a checkout
```dart
CheckoutObject checkout= await coinbase.deleteCheckout('YOUR CHECKOUT ID HERE');
```

#### To List all charges assoiciated with an account

```dart
List<CheckoutObject> checkoutList = await coinbase.retrieveAllCheckouts;
```

### Invoices

#### To create a new invoice
```dart
InvoiceObject invoice = await coinbase.createInvoice(
                      businessName: 'businessName',
                      customerEmail: 'onuohasilver9@gmail.com',
                      customerName: 'Onuoha Silver',
                      currency: CurrencyType.usd,
                      amount: 20,
                    );
```

#### To view an invoice
```dart
InvoiceObject invoice = await coinbase
                        .viewInvoice('2a4aeb3e-a12b-4728-9c88-297dea5d2adb');
```
#### To retrieve all Invoices
```dart
List<InvoiceObject> invoice =await coinbase.retrieveAllInvoices;
```











CHANGELOG.md


## [0.0.1] - 2nd May 2021.

* First Release.



## [0.0.2] - 2nd May 2021.

* Add null safety


## [0.0.3] - 28th May 2021.

* Add Simple Documentations
* Commenting and fixing ChargeObject bugs


example/lib/main.dart
import 'package:coinbase_commerce/coinbase_commerce.dart';
import 'package:coinbase_commerce/returnObjects/statusObject.dart';

import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: ExampleScreen(),
    );
  }
}

class ExampleScreen extends StatefulWidget {
  @override
  _ExampleScreenState createState() => _ExampleScreenState();
}

String apiKey = '';
Coinbase coinbase = Coinbase(apiKey, debug: true);
String output = 'Output';

class _ExampleScreenState extends State<ExampleScreen> {
  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    return Material(
      child: Container(
        height: size.height,
        width: size.width,
        color: Colors.white,
        child: Column(
          children: [
            SizedBox(height: 50),
            Text(
              'Coinbase Example App',
              style: TextStyle(fontSize: 30),
            ),
            Wrap(
              children: [
                TextButton(
                  onPressed: () async {
                    ChargeObject image = await coinbase.createCharge(
                        name: 'Record Me',
                        description: 'description',
                        currency: CurrencyType.usd,
                        pricingType: PricingType.fixedPrice,
                        amount: 40);

                    StatusObject status = coinbase.checkChargeStatus(image);
                    setState(() {
                      output = status.status.toString();
                    });
                  },
                  child: Text('Create a Charge'),
                ),
                TextButton(
                  onPressed: () async {
                    ChargeObject image =
                        await coinbase.cancelCharge('MJ7ED8BQ');
                    StatusObject status = coinbase.checkChargeStatus(image);
                    setState(() {
                      output = status.status.toString();
                    });
                  },
                  child: Text('Cancel a Charge'),
                ),
                TextButton(
                  onPressed: () async {
                    ChargeObject image = await coinbase.viewCharge('RLWADR5Z');
                    setState(() {
                      output = image.toString();
                    });
                  },
                  child: Text('View a Charge'),
                ),
                TextButton(
                  onPressed: () async {
                    CheckoutObject image = await coinbase.createCheckout(
                        description: 'description',
                        name: 'name',
                        pricingType: PricingType.noPrice,
                        // amount: 30,
                        currency: CurrencyType.usd);

                    setState(() {
                      output = image.toString();
                    });
                  },
                  child: Text('Create a Checkout'),
                ),
                TextButton(
                  onPressed: () async {
                    CheckoutObject image = await coinbase.updateCheckout(
                        checkoutID: 'c4357d83-42d4-4d7b-bc44-d0262b1efd25',
                        description: 'description',
                        name: 'Silver Gold',
                        pricingType: PricingType.noPrice,
                        currency: CurrencyType.usd);
                    setState(() {
                      output = image.toString();
                    });
                  },
                  child: Text('Update a Checkout'),
                ),
                TextButton(
                  onPressed: () async {
                    CheckoutObject image = await coinbase
                        .viewCheckout('c4357d83-42d4-4d7b-bc44-d0262b1efd25');
                    setState(() {
                      output = image.toString();
                    });
                  },
                  child: Text('View  Checkout'),
                ),
                TextButton(
                  onPressed: () async {
                    CheckoutObject image = await coinbase.deleteCheckout(
                        checkoutID: 'c4357d83-42d4-4d7b-bc44-d0262b1efd25');
                    setState(() {
                      output = image.toString();
                    });
                  },
                  child: Text('Delete checkout'),
                ),
                TextButton(
                  onPressed: () async {
                    InvoiceObject image = await coinbase.createInvoice(
                      businessName: 'businessName',
                      customerEmail: 'onuohasilver9@gmail.com',
                      customerName: 'Onuoha Silver',
                      currency: CurrencyType.usd,
                      amount: 20,
                    );
                    setState(() {
                      output = image.toString();
                    });
                  },
                  child: Text('Create an Invoice'),
                ),
                TextButton(
                  onPressed: () async {
                    InvoiceObject image =
                        await coinbase.voidInvoice(code: 'ZY259PAT');
                    setState(() {
                      output = image.toString();
                    });
                  },
                  child: Text('Void an Invoice'),
                ),
                TextButton(
                  onPressed: () async {
                    InvoiceObject image = await coinbase
                        .viewInvoice('2a4aeb3e-a12b-4728-9c88-297dea5d2adb');
                    setState(() {
                      output = image.toString();
                    });
                  },
                  child: Text('View an Invoice'),
                ),
                TextButton(
                  onPressed: () async {
                    List<ChargeObject> image =
                        await coinbase.retrieveAllCharges;
                    setState(() {
                      output = image.length.toString();
                    });
                  },
                  child: Text('List all charges'),
                ),
                TextButton(
                  onPressed: () async {
                    List<CheckoutObject> image =
                        await coinbase.retrieveAllCheckouts;

                    setState(() {
                      output = image.length.toString();
                    });
                  },
                  child: Text('List all checkouts'),
                ),
                TextButton(
                  onPressed: () async {
                    List<InvoiceObject> image =
                        await coinbase.retrieveAllInvoices;
                    setState(() {
                      output = image.length.toString();
                    });
                  },
                  child: Text('List all invoices'),
                ),
              ],
            ),
            Text(
              output,
              style: TextStyle(fontSize: 20),
            )
          ],
        ),
      ),
    );
  }
}
